path_login.go

```diff
@@ -52,7 +79,21 @@ has included a signature.`,
 // to this function receive logical errors instead of internal server errors where appropriate
 // logic updates relating to role determination should be kept consistent between the two.
 func (b *backend) pathLoginResolveRole(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
-	b64URL := data.Get("identity_request_url").(string)
+	var err error
```

Comment:
Hmm, I wonder if `pathLoginResolveRole()` could be doing too much? 🤔 I was looking at [vault-plugin-auth-jwt/path_login.go#L48](https://github.com/hashicorp/vault-plugin-auth-jwt/blob/main/path_login.go#L48) as a reference and this seems to be doing quite a bit more.

The reason I ask is somewhat related to my other comment around moving the `tools.GenerateLoginData()` call from the Vault CLI to Vault server. I'm wondering if we could avoid doing that and avoid the nonce reuse issue by doing less in `pathLoginResolveRole()`. I'm not exactly sure where the nonce reuse issue surfaced, but perhaps we could remove the call to `b.getCallerIdentity()` when resolving the role?

Reply:
I had a similar thought during implementation. But then I realized that the code here doesn't require the role to be set. So if it isn't set, we make the GetCallerIdentity request and use the role that we parse from the returned ARN.

I am just now noticing that our docs do say that the [role is required on login](https://developer.hashicorp.com/vault/api-docs/auth/alicloud#role-3). I wonder if it is reasonable to now require the role? This could allow for us to simplify `pathLoginResolveRole` similar to auth-jwt (which doesn't require the role but it has a default).

Reply:
I have thought about it some more and I don't see any other path forward except to make the `role` a required field on login. Given that the docs already say it is required, I think it is the best choice. It would also require the least amount of new code and we could greatly simplify resolveRole as you said. Let me know if you agree/disagree here. Thanks!

Reply:
That sounds good to me! Looks like it's required for [agent auto-auth](https://developer.hashicorp.com/vault/docs/agent/autoauth/methods/alicloud) too. I'm good going forward with that 👍